### PR TITLE
Unit tests restructuring

### PR DESCRIPTION
**Pull request by nikithauc**
This PR focusses to improve the existing test structure of the project. Changes are as follows -
1. Added testing tools -
		a. `Karma` - To test the unit test in browsers
		b. `@istanbuljs/nyc-config-typescript` - Evaluation of the test coverage
2. Restructured the test folders as follows 
		a. Renamed from folder spec/  to test/
		b. Divided unit test in folders - common, node and browser.
		c. Removed the config files from the test folder.
3. The JS SDK is transpiled into CommonJs and ESM modules. Added `esm` dependency to test run mocha tests on transpiled ESM modules.
4. Some tests were failing when I was using the latest version of isomorphic-fetch/ cross-fetch/via karma. Updated those  tests to align to the updates in the `fetch api`. Updated  the isomorphic-fetch dependency. 
5. Corrected some invalid tests.   
6. Added headers as a property in the ChaosHandlerOptions. I plan for some more the ChaosHandlerOptions design changes but not in this PR.

TODO - Check if the any changes are needed in the GraphResponseHandler to adhere to fetch api/http/ Graph Response standards. Will create a follow up PR if any changes are present

### CODE REVIEW

**Review by baywet — APPROVED** [review]
Great work, this modernizes the test infrastructure. I noticed some TODO's were remaining, but my guess is that it's why this PR is still in draft.

**Comment by nikithauc** [comment]
> Great work, this modernizes the test infrastructure. I noticed some TODO's were remaining, but my guess is that it's why this PR is still in draft.

Thanks! This PR is not marked as draft. It is #371  that is marked as draft.

**Comment by baywet** [comment]
For reference here are the TODO I was talking about https://github.com/microsoftgraph/msgraph-sdk-javascript/pull/370/files#diff-d4ad0dac051d781735e98ebecb49ec8c172723a92f37910c1c413c414dd2df23

File: test/node/NodeTests.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+This folder contains unit tests which are node-specific.
+
+Karma is used for testing the code in browsers. To run the unit tests using karma run the following steps -
+
+-   Run the script npm install
```
Comment (MIchaelMainer) [inline test/node/NodeTests.md:5]: nit 

Run the script `npm install`
Run the script `npm run test`

File: test/node/content/BatchResponseContent.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		it("Should update instance with nextLink responses", (done) => {
+			const responses = [redirect, forbidden];
+			const batchRes = new BatchResponseContent({
+				responses,
+				"@odata.nextLink": "dummy nextlink",
+			});
+			batchRes.update({
+				responses: [okWithoutBody, okWithBody],
+			});
+			assert.isDefined(batchRes.getResponseById("2"));
```
Comment (MIchaelMainer) [inline test/node/content/BatchResponseContent.ts:56]: I suggest that we test that the BatchResponseContent.Update doesn't delete/alter the existing batch responses. Might as well:

```javascript
assert.isDefined(batchRes.getResponseById("1"));
assert.isDefined(batchRes.getResponseById("2"));
assert.isDefined(batchRes.getResponseById("3"));
assert.isDefined(batchRes.getResponseById("4"));
```

This does lead me to wonder why `update()` is public. I don't think we want want customers adding responses to BatchResponseContent.

**Review by MIchaelMainer — APPROVED** [review]
with comments

File: test/node/NodeTests.md (hunk 1 above)
Comment (nikithauc) [reply test/node/NodeTests.md]: Done

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -114,10 +114,9 @@ npm install
 
 To edit files, open them in an editor of your choice and modify them. To create a new file, use the editor of your choice and save the new file in the appropriate location in your local copy of the repository. While working, save your work frequently.
 
-*Note: Make sure to add unit tests to validate you changes.*
+_Note: Make sure to add unit tests to validate you changes._
 
-Once you have done with your changes, You have to build and test your changes
-To build the library run,
+Once you have done with your changes, You have to build and test your changes To build the library run,
```
[the remark this answers is not in the document]
Comment (nikithauc) [reply CONTRIBUTING.md]: Done

File: test/node/content/BatchResponseContent.ts (hunk 2 above)
Comment (nikithauc) [reply test/node/content/BatchResponseContent.ts]: Done

**Pull request merged**